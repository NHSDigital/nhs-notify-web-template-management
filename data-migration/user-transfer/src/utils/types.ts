export type Parameters = {
  environment: string;
  component: string;
  accessKeyId: string;
  secretAccessKey: string;
  userPoolId: string;
  region: string;
  sessionToken: string;
  flag?: string;
};

export type UserData = {
  username: string;
  /*
   * extracted from the user's group starting with 'client:'
   */
  clientId: string;
  /*
   * extracted from the user's 'sub' attribute
   */
  userId: string;
};

export type Template = {
  id: string;
  owner: string;
};

type ToFrom = {
  from: string;
  to: string;
};

export type TransferStage =
  | 's3:copy'
  | 'ddb:transfer'
  | 's3:delete'
  | 'initial'
  | 'finished';

export type TransferStatus = 'success' | 'failed' | 'migrate';

export type UserTransferPlanItem = {
  templateId: string;
  status: TransferStatus;
  stage: TransferStage;
  reason?: string;
  ddb: {
    owner: ToFrom;
  };
  s3: {
    files: ToFrom[];
  };
};

export type UserTransferPlan = {
  total: number;
  tableName: string;
  bucketName: string;
  migrate: {
    count: number;
    plans: UserTransferPlanItem[];
  };
  orphaned: {
    count: number;
    templateIds: string[];
  };
};

export type UserTransferPlanItemResult = {
  success: boolean;
  stage: TransferStage;
  reasons?: string[];
};
