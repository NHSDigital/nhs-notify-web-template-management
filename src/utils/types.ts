export type FormId =
  | 'choose-a-template-type'
  | 'create-nhs-app-template'
  | 'create-email-template'
  | 'create-letter-template'
  | 'create-text-message-template'
  | 'preview-nhs-app-template'
  | 'submit-template'
  | 'create-nhs-app-template-back';

export type FormErrorState = {
  formErrors: string[];
  fieldErrors: Record<string, string[]>;
};

export enum TemplateType {
  NHS_APP = 'NHS_APP',
  SMS = 'SMS',
  EMAIL = 'EMAIL',
  LETTER = 'LETTER',
}

export type Session = {
  id: string;
  templateType: TemplateType | 'UNKNOWN';
  nhsAppTemplateName: string;
  nhsAppTemplateMessage: string;
};

export type FormState = {
  validationError?: FormErrorState;
  redirect?: string;
};

export type TemplateFormState = FormState & Session;

export type PageProps = {
  params: {
    sessionId: string;
  };
};

export type TemplateSubmittedPageProps = {
  params: {
    templateId: string;
  };
};

export type PageComponentProps = {
  initialState: TemplateFormState;
};

export type SubmitTemplatePageComponentProps = {
  templateName: string;
  sessionId: string;
};

export enum TemplateFormatText {
  APP = 'APP',
  EMAIL = 'EMAIL',
  LETTER = 'LETTER',
  SMS = 'SMS',
}
