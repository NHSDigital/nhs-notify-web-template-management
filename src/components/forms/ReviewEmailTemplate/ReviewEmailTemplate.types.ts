export type ReviewEmailTemplateProps = {
  templateName: string;
  subject: string;
  message: string;
};
