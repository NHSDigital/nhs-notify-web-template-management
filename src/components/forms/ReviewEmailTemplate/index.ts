export * from './ReviewEmailTemplate';
export * from './ReviewEmailTemplate.types';
export * from './server-actions';
