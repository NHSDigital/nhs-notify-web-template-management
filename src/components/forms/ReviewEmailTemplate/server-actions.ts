import { FormState } from '@utils/types';
import { MarkdownItWrapper } from '@utils/markdownit';

export function renderMarkdown(
  value: string,
  markdown = new MarkdownItWrapper()
) {
  markdown.enableLineBreak().enable(['heading', 'link', 'list', 'hr']);

  return markdown.render(value);
}

export function handleForm(state: FormState, _data: FormData) {
  return { ...state };
}
