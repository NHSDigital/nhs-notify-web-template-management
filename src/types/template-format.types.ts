export enum TemplateFormatText {
  APP = 'APP',
  EMAIL = 'EMAIL',
  LETTER = 'LETTER',
  SMS = 'SMS',
}
