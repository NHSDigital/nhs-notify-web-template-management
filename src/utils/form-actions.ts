'use server';

import { getAmplifyBackendClient } from '@utils/amplify-utils';
import { DbOperationError } from '@domain/errors';
import { Template, Draft } from './types';
import { logger } from './logger';

export async function createTemplate(
  template: Draft<Template>
): Promise<Template> {
  const { data, errors } =
    await getAmplifyBackendClient().models.TemplateStorage.create(template);

  if (errors || !data) {
    logger.error('Failed to create template', errors);
    throw new Error('Failed to create new template');
  }
  return data;
}

export async function saveTemplate(template: Template): Promise<Template> {
  const { data, errors } =
    await getAmplifyBackendClient().models.TemplateStorage.update(template);

  if (errors) {
    logger.error('Failed to save template', errors);
    throw new Error('Failed to save template data');
  }

  if (!data) {
    throw new DbOperationError({
      message:
        'Template in unknown state. No errors reported but entity returned as falsy',
      operation: 'update',
    });
  }
  return data;
}

export async function getTemplate(
  templateId: string
): Promise<Template | undefined> {
  const { data, errors } =
    await getAmplifyBackendClient().models.TemplateStorage.get({
      id: templateId,
    });

  if (errors) {
    logger.error('Failed to get template', errors);
  }

  if (!data) {
    logger.warn(`Failed to retrieve template for ID ${templateId}`);
    return undefined;
  }

  return data;
}

export async function sendEmail(
  templateId: string,
  templateName: string,
  templateMessage: string,
  templateSubjectLine: string | null
) {
  const res = await getAmplifyBackendClient().queries.sendEmail({
    recipientEmail: 'england.test.cm@nhs.net',
    templateId,
    templateName,
    templateMessage,
    templateSubjectLine,
  });

  if (res.errors) {
    logger.error({
      description: 'Error sending email',
      res,
    });
  }
}

export async function getTemplates(): Promise<Template[] | []> {
  const { data, errors } =
    await getAmplifyBackendClient().models.TemplateStorage.list();

  if (errors) {
    logger.error('Failed to get templates', errors);
  }

  if (!data) {
    logger.warn(`Failed to retrieve templates`);
    return [];
  }

  return data;
}
