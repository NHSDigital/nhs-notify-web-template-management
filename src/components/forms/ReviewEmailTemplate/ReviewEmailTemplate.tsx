'use client';

import { PreviewTemplate } from '@molecules/PreviewTemplate';
import { ReviewTemplate } from '@organisms/ReviewTemplate';
import content from '@/src/content/content';
import { renderMarkdown } from './server-actions';

export type ReviewEmailTemplateProps = {
  templateName: string;
  subject: string;
  message: string;
};

export function ReviewEmailTemplate({
  templateName,
  subject,
  message,
}: Readonly<ReviewEmailTemplateProps>) {
  const {
    components: {
      reviewEmailTemplateContent: { sectionHeading, details, form },
    },
  } = content;

  const html = renderMarkdown(message);

  return (
    <div className='nhsuk-grid-row'>
      <div className='nhsuk-grid-column-two-thirds'>
        <ReviewTemplate
          templateName={templateName}
          sectionHeading={sectionHeading}
          details={details}
          form={{
            ...form,
            action: '',
            state: {
              validationError: undefined,
            },
            formId: 'review-email-template',
            radiosId: 'reviewEmailTemplateAction',
          }}
          PreviewComponent={
            <PreviewTemplate.Email subject={subject} value={html} />
          }
        />
      </div>
    </div>
  );
}
