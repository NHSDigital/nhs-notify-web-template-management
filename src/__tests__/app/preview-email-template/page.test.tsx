import PreviewEmailTemplatePage from '@app/preview-email-template/[sessionId]/page';
import { ReviewEmailTemplate } from '@forms/ReviewEmailTemplate';
import { TemplateType } from '@utils/types';
import { redirect } from 'next/navigation';
import { getSession } from '@utils/form-actions';

jest.mock('@utils/form-actions');
jest.mock('next/navigation');
jest.mock('@forms/ReviewEmailTemplate');

const redirectMock = jest.mocked(redirect);
const getSessionMock = jest.mocked(getSession);

describe('PreviewEmailTemplatePage', () => {
  beforeEach(jest.resetAllMocks);

  it('should load page', async () => {
    const state = {
      id: 'session-id',
      templateType: TemplateType.EMAIL,
      emailTemplateName: 'template-name',
      emailTemplateSubjectLine: 'template-subject-line',
      emailTemplateMessage: 'template-message',
      nhsAppTemplateMessage: '',
      nhsAppTemplateName: '',
    };

    getSessionMock.mockResolvedValueOnce(state);

    const page = await PreviewEmailTemplatePage({
      params: {
        sessionId: 'session-id',
      },
    });

    expect(page).toEqual(<ReviewEmailTemplate initialState={state}/>);
  });

  it('should redirect to invalid-session when no session is found', async () => {
    await PreviewEmailTemplatePage({
      params: {
        sessionId: 'session-id',
      },
    });

    expect(redirectMock).toHaveBeenCalledWith('/invalid-session', 'replace');
  });

  test.each([
    {
      templateType: TemplateType.LETTER,
      emailTemplateName: 'valid-name',
      emailTemplateSubjectLine: 'valid-subject-line',
      emailTemplateMessage: 'valid-message',
    },
    {
      templateType: TemplateType.SMS,
      emailTemplateName: 'valid-name',
      emailTemplateSubjectLine: 'valid-subject-line',
      emailTemplateMessage: 'valid-message',
    },
    {
      templateType: TemplateType.NHS_APP,
      emailTemplateName: 'valid-name',
      emailTemplateSubjectLine: 'valid-subject-line',
      emailTemplateMessage: 'valid-message',
    },
    {
      templateType: TemplateType.EMAIL,
      emailTemplateName: undefined,
      emailTemplateSubjectLine: 'valid-subject-line',
      emailTemplateMessage: 'valid-message',
    },
    {
      templateType: TemplateType.EMAIL,
      emailTemplateName: 'valid-name',
      emailTemplateSubjectLine: undefined,
      emailTemplateMessage: 'valid-message',
    },
    {
      templateType: TemplateType.EMAIL,
      emailTemplateName: 'valid-name',
      emailTemplateSubjectLine: 'valid-subject-line',
      emailTemplateMessage: undefined,
    },
    {
      templateType: TemplateType.EMAIL,
      emailTemplateName: 'valid-name',
      emailTemplateSubjectLine: 'valid-subject-line',
      emailTemplateMessage: null as unknown as string,
    },
  ])(
    'should redirect to invalid-session when session template is $templateType and name is $emailTemplateName and subjectLine is $$emailTemplateSubjectLine and message is $emailTemplateMessage',
    async (value) => {
      getSessionMock.mockResolvedValueOnce({
        id: 'session-id',
        nhsAppTemplateMessage: '',
        nhsAppTemplateName: '',
        ...value,
      });

      await PreviewEmailTemplatePage({
        params: {
          sessionId: 'session-id',
        },
      });

      expect(redirectMock).toHaveBeenCalledWith('/invalid-session', 'replace');
    }
  );
});
