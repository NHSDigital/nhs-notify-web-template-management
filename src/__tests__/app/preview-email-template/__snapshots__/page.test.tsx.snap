// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PreviewEmailTemplatePage should render ReviewEmailTemplate with session data 1`] = `
<ReviewEmailTemplate
  initialState={
    {
      "emailTemplateMessage": "template-message",
      "emailTemplateName": "template-name",
      "emailTemplateSubjectLine": "template-subject-line",
      "id": "session-id",
      "nhsAppTemplateMessage": "",
      "nhsAppTemplateName": "",
      "templateType": "EMAIL",
    }
  }
/>
`;
